### DISCUSSION TITLE
Bindings not working using ObservableProperty

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#17440] [Q&A] [answered]
Hi
I have gone through some docs and google help and trying to get my View updated when the variables in UserOptions change.   How can I bind LiveCount to my TextBox using xaml in my View?  At the moment it's just empty
If I just define a single variable in the MainViewModel just to test it seems to work, but when using an object class it doesnt

e.g. the binding of Timenow works
  ```
[ObservableProperty]
  private string _timenow;
```

```
public partial class MainViewModel : ObservableObject
{
    private INavigator _navigator;
    public UserOptions useroptions;
.....
}
```

the useroptions gets updated here by connecting and getting the payload from my signalR hub

```
private async Task<UserOptions> GetUserPreferences(HubConnection connection)
{   
    try
    {
        useroptions = await connection.InvokeAsync<UserOptions>("GetUserPreferences");
    }
    catch (Exception ex)
    { }
}
```

my Model class is below (`UserOptions.cs`) (LiveCount doesnt bind):
```
public partial class UserOptions : ObservableObject
{
    [ObservableProperty]
    private int? liveCount;
} 
```
MainPage.xaml.cs

```

using Microsoft.UI.Windowing;

namespace AirMyPrayerMenu.Presentation;

public sealed partial class MainPage : Page
{
    private MainViewModel ViewModel => (MainViewModel)DataContext;
   
    public MainPage()
    {
        this.InitializeComponent();
    }

    private void Page_DataContextChanged(FrameworkElement sender, DataContextChangedEventArgs args)
    {
        var presenter = GetPresenter();

    }

    private OverlappedPresenter GetPresenter()
    {
        var app = Application.Current as App;
        var presenter = app.MainWindow.AppWindow.Presenter;
        app.MainWindow.Title = "MyTitle";   
        return (OverlappedPresenter)presenter;
    }
} 

```

I can see the useroptions.LiveCount when debugging, the field has been populated but my binding shows nothing

`<TextBlock HorizontalAlignment="Center" Margin="0,0,0,0">Live Users: <Run Text="{Binding LiveCount}" /></TextBlock>`

**Comment by morning4coffe-dev** [comment]
Hey @ashah7, if I understood your question correctly you want to use the LiveCount that is in the UserOptions in your MainPage, right?

For this you can make your UserOptions a property in the MainViewModel ```public UserOptions UserOptions { get; }``` and when there is an instance of the UserOptions you can assign it in the MainPage.xaml like this: 
```
<TextBlock HorizontalAlignment="Center" Margin="0,0,0,0">Live Users: <Run Text="{Binding UserOptions.LiveCount}" /></TextBlock>
```

**Reply by ashah7** [reply]
Hi
that sound correct, unfortunately I have just tried but its still not binding

```
public partial class MainViewModel : ObservableObject
{

    private INavigator _navigator;
    public UserOptions UserOptions { get; }
    public UserOptions useroptions ;
}
````
` <TextBlock HorizontalAlignment="Center" Margin="0,0,0,0">Live Users: <Run Text="{Binding UserOptions.LiveCount}" /></TextBlock>`

**Reply by ashah7** [reply]
Just to clarify my UserOptions model is created under the "Models" folder as `UserOptions.cs`

**Reply by morning4coffe-dev** [reply, marked as answer]
I see, did you assign the UserOptions property with a value from your GetUserPreferences? In you case I think you can do something like this:
MainViewModel.cs:
```cs
    [ObservableProperty]
    private UserOptions userOptions;

    public MainViewModel()
    {
        //You can assign it from your GetUserPreferences later as well
        UserOptions = new UserOptions();

        UpdateButtonCommand = new RelayCommand(UpdateButton);
    }
    public ICommand UpdateButtonCommand { get; }

    //The Update Button just to show the value update of LiveCount
    private void UpdateButton()
    {
        UserOptions.LiveCount++;
    }

    //private async Task<UserOptions> GetUserPreferences(HubConnection connection)
    //{
    //    try
    //    {
    //        UserOptions = await connection.InvokeAsync<UserOptions>("GetUserPreferences");
    //    }
    //    catch (Exception ex)
    //    { }
    //}
```

UserOptions.cs:
```cs
    [ObservableProperty]
    private int? liveCount = 0;
```

MainPage.xaml:
```xml
    <TextBlock HorizontalAlignment="Center"
               Margin="0,0,0,0">Live Users: <Run Text="{Binding UserOptions.LiveCount}" /></TextBlock>
    <Button Content="Update"
            Command="{Binding UpdateButtonCommand}" />
```


https://github.com/unoplatform/uno/assets/78549750/400b07b4-c5bc-49c4-893b-0b38bb9606be


I added the Update Button just to show the update of the property, you don't need to use it.

**Reply by ashah7** [reply]
thanks that looks like it worked. Struggled with this for days, appreciated.
It looks like this line did it
`UserOptions = new UserOptions();`

not quite sure what was the problem with the below and using `useroptions`

`public UserOptions useroptions;`

**Reply by morning4coffe-dev** [reply]
Awesome, glad to hear it worked. Jup, the reason was that the `UserOptions` property didn't have any value assigned to itself. For binding in XAML you should use properties instead of fields (`useroptions`).
